File: ra-oauth2-idp/src/ports/key-store.ts
/**
 * Layer 3 — Application Logic（ポート定義）
 *
 * 署名鍵ストア。本番では KMS / HSM、本アプリではメモリ。
 * 公開鍵は JWKS として配布される。
 *
 * 鍵オブジェクト本体（`privateKey` / `publicKey`）の表現は実行環境依存
 * （Web Crypto の `CryptoKey`、Node の `KeyObject`、HSM ハンドル等）。
 * ポート層では「不透明な鍵参照」として `unknown` 扱いとし、
 * 署名アダプター層が自分の知っている形に narrowing する。
 */

export interface SigningKey {
  kid: string
  alg: 'PS256' | 'ES256'
  privateKey: unknown
  publicKey: unknown
  publicJwk: Record<string, unknown>
  active: boolean
  created_at: string
}

export interface KeyStore {
  /** 現在アクティブな署名鍵（1つ）。 */
  getActiveKey(): Promise<SigningKey>
  /** 旧鍵も含めた全鍵（JWKS で公開する用途）。 */
  getAllKeys(): Promise<SigningKey[]>
  /** 指定 kid の鍵（検証用途）。 */
  findByKid(kid: string): Promise<SigningKey | null>
  /** 新しい鍵に回転する。 */
  rotate(): Promise<SigningKey>
}

File: ra-idp-go/internal/oauth2/ports/key_store.go
package ports

import (
	"context"
	"crypto"
	"time"

	"ra-idp-go/internal/spec"
)

// SigningKey は本実装では RSA を想定。alg=PS256 のみ。
// 公開鍵 JWK は JWKS 配布用。
type SigningKey struct {
	Kid        string
	Alg        spec.SignatureAlgorithm
	PrivateKey crypto.PrivateKey
	PublicKey  crypto.PublicKey
	PublicJWK  map[string]any
	Active     bool
	CreatedAt  time.Time
}

type KeyStore interface {
	GetActiveKey(ctx context.Context) (*SigningKey, error)
	GetAllKeys(ctx context.Context) ([]*SigningKey, error)
	FindByKID(ctx context.Context, kid string) (*SigningKey, error)
	Rotate(ctx context.Context) (*SigningKey, error)
}
